### ISSUE TITLE
pmesh with dynamic box size

### ISSUE THREAD

**Issue by mortele** [rainwoodman/pmesh#42]
We have been trying to implement a scheme for varying the simulation box *on the fly* using pmesh. We have some ideas on how to do this, but we are not sure which way is best (or even feasible). 

In our implementation, the box will change every time step or every `N` time steps (albeit most of the time very slightly).

First attempt
-------------
Our first approach was just to re-create the entire pmesh machinery every time the box changes. So in the molecular dynamics (MD) loop, we do something akin to 
```python
box_size = ...
pm = pmesh.ParticleMesh(mesh_size, BoxSize=box_size)
real_field = pm.create("real", value=0.0)
complex_field = pm.create("complex", value=0.0)

for t in time_steps:
   # compute forces
   # update positions and velocities
   dbox = compute_box_change(...)
   box_size = box_size + dbox
   
   # re-create the pmesh object and any dependents with the new box size
   pm = pmesh.ParticleMesh(mesh_size, BoxSize=box_size)
   real_field = pm.create("real", value=0.0)
   complex_field = pm.create("complex", value=0.0)
```
This is somewhat of a drastic solution, but since our problem has no *memory* (of past states) and can only depend on the current positions/velocities of the particles, this is at least guaranteed to be correct. However, we expect it to be slow, having to re-do the setup and allocate all field arrays every time step.

Second attempt
----------------
Our second try was to inspect the pmesh objects and see if we could decipher which internals depend explicitly on the box size, and which would be unaffected, and then simply changing those attributes by hand. We guessed that this would be *much* faster.
```python
>>> import pmesh.pm as pmesh
>>> pm = pmesh.ParticleMesh([5,6,7], BoxSize=[10.0, 12.0, 14.0], dtype="f4")
>>> real_field = pm.create("real", value=0.0)
>>> complex_field = pm.create("complex", value=0.0)
```
Inspecting the `pm` object reveals some quantities which are obviously dependent on the box size, and some which are more subtle:
```python
>>> for k, v in pm.__dict__.items():
...     print(k, v)
...
# lightly edited, some output omitted
BoxSize [10. 12. 14.]
domain <pmesh.domain.GridND object at 0x10be01280>
procmesh <pfft.core.ProcMesh object at 0x10be70400>
affine <pmesh.window.Affine object at 0x10be012b0>
affine_grid <pmesh.window.Affine object at 0x10be01340>
resampler <pmesh.window.ResampleWindow object at 0x7fbd20181010>
template <pmesh.pm._pmtemplate object at 0x10be01310>
_coords {
   <class 'pmesh.pm.RealField'>: ([array(...)]),
   <class 'pmesh.pm.TransposedComplexField'>: ([array(...)])
}
```
Obviously, the `pm.BoxSize` variable depends explicitly on the size of the box. In addition, the `pm.affine` child has a `scale` attribute which appears to be `Nmesh / BoxSize`. Also, fields in `pm._coords` change when the box changes.

Otherwise, to us it looks like `domain`, `procmesh`, `affine`, `affine_grid`, `resampler`, and `template` all remain unaltered.

In the same way, for the fields we find
```python
>>> for k, v in real_field.__dict__.items():
...     print(k, v)
...
# lightly edited, some output omitted
_base <pfft.core.LocalBuffer object at 0x10be70bd0>
pm <pmesh.pm.ParticleMesh object at 0x10bcda040>
BoxSize [10. 12. 14.]
value [[[...]]]
x [array([[[ 0.]],
          [[ 2.]],
          ...
]
i [array([[[ 0]],
          [[ 1]],
          ...
]
```
of which `BoxSize` and `x` both depend explicitly on the box size, and `pm` points to the parent object which we intend to change. It is unclear if the `pfft.core.LocalBuffer` object in `real_field._base` has any internals dependent on the box size, the cython code is harder to introspect. 

#### Why did we care about all these objects, again
Our thought was that we could now&mdash;instead of re-creating the whole machinery&mdash;just manually set the attributes of `pm` and the fields to the new values corresponding to the new box size, and then continue running the MD loop. Thus, hopefully, saving a lot of time in the code.

### Some preliminary questions
- Do you think this second approach is at all feasible?
- Did we miss some obvious dependencies on the box size somewhere in the internals of pmesh or pfft-python?
- You mentioned trying to define a "comoving" coordinate system with the box size internal in pmesh always being e.g. `[1, 1, 1]` and handling the transformation to/from local (internal) ↔ global coordinate system ourselves in the code; I think if the second approach turns out to be unfruitful, we should look into how to do this.

**Comment by mortele** [comment]
Maybe you can add some brief background and equations on what specifically we are trying to do @samiransen23 ?

**Comment by samiransen23** [comment]
I think you explained the problem pretty well. The physics we are dealing with is as follows:
We use the Berendsen barostat to compute a factor &alpha; to rescale the box size and the positions (x,y,z) of the particles.  
<a href="https://www.codecogs.com/eqnedit.php?latex=L_{box}&space;=&space;\alpha^{1/3}&space;L_{box}&space;\\&space;(x,y,z)&space;=&space;\alpha^{1/3}&space;(x,y,z)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?L_{box}&space;=&space;\alpha^{1/3}&space;L_{box}&space;\\&space;(x,y,z)&space;=&space;\alpha^{1/3}&space;(x,y,z)" title="L_{box} = \alpha^{1/3} L_{box} \\ (x,y,z) = \alpha^{1/3} (x,y,z)" /></a>

To continue running the MD, we use several fields derived from `pm`. But if the box size changes, then the foundation i.e. the grid drawn through the box space changes i.e. `pm` changes. So we recreate `pm` and update all those fields' attributes that we "think" is affected by box-size. But we are not sure about some of the attributes as Morten said above.   
Currently, all I do whenever the box and positions are rescaled is:
```python
pm = pmesh.ParticleMesh(config.mesh_size, BoxSize=config.box_size, dtype="f4", comm=comm)
for f in field_list:
    f.pm = pm
    f.BoxSize = pm.BoxSize 
```
`field_list` contains all fields created using `pm.create`.  
Rescaled positions are painted onto the new grid by `pm.paint(positions, ...)`. 

**The main question is:** If the box size changes, what all do we need to change/recreate to update `pm` and all fields derived from it?

**Comment by samiransen23** [comment]
See #44

**Comment by rainwoodman** [comment]
What did you do in the end? Would you file a PR to improve the caching mechanism for different box sizes?

If not I can take a stab.

**Comment by samiransen23** [comment]
Hey, I've commented it in #44 because it seemed to cleaner there. It was not a big change, so a PR wouldn't be necessary. Thanks 😊

**Comment by rainwoodman** [comment]
So the answer to this question is, yes, if the boxsize changes we do need a new pm object. The pm object is dimension-aware.

Alternatively, you can define your field on a dimensionless coordinate system, such that the boxsize pm object sees is always 1.0?

**Comment by rainwoodman** [comment]
Closed by 174b397. Thanks!
